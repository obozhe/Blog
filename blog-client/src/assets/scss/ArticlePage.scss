.article-page {
  padding: 1em;
  min-height: 100vh;

  .post-wrapper {
    min-width: 500px;
    width: fit-content;
    margin: 2em auto 2em auto;
    .post {
      background-color: $white;
      padding: 2em;
      border-radius: 10px;
      box-shadow: 0 0 20px -18px #050505;

      .date {
        float: right;
      }

      .date,
      .author {
        opacity: 0.4;
      }

      .title {
        width: 100%;
        text-align: center;
      }
    }
  }
  .arrows {
    position: relative;
    background-color: rgba($white, 0.3);
    margin-bottom: 10px;
    padding: 0.5em 1em;
    border-radius: 5px;
    display: flex;
    align-items: center;
    justify-content: space-between;

    a {
      text-decoration: none;
      color: inherit;
    }
    
    .disabled-link {
      pointer-events: none;
      opacity: 0.3;
    }
  }
}
