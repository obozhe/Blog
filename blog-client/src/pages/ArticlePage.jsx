import React from 'react';
import IconButton from '@material-ui/core/IconButton';
import ArrowBack from '@material-ui/icons/ArrowBack';
import ArrowForward from '@material-ui/icons/ArrowForward';
import Edit from '@material-ui/icons/Edit';
import Delete from '@material-ui/icons/Delete';
import { Link } from 'react-router-dom';
import AddPost from '../components/AddPost';
import Loader from '../components/Loader';

export default class ArticlePage extends React.Component {
  constructor() {
    super();
    this.state = {
      post: null,
      openEditModal: false,
      loading: true,
    };
  }

  componentDidMount() {
    this.setPost();
  }

  setPost() {
    const { getPost, id } = this.props;
    getPost(id).then((post) => {
      this.setState({ post, loading: false });
    });
  }

  render() {
    const {
      id, extraIds, deletePost, editPost,
    } = this.props;

    const { loading, openEditModal, post } = this.state;
    return (
      <div className="article-page">
        <Link to="/">
          <IconButton aria-label="Back to main">
            <ArrowBack fontSize="small" />
          </IconButton>
        </Link>
        {!loading ? (
          <div className="post-wrapper">
            <div className="arrows">
              <Link
                to={`/${extraIds.prev}`}
                className={extraIds.prev ? 'prev' : 'disabled-link prev'}
              >
                <ArrowBack fontSize="small" />
              </Link>
              <span>
                <IconButton
                  aria-label="edit post"
                  onClick={() => this.setState({ openEditModal: true })}
                >
                  <Edit color="primary" />
                </IconButton>
                <IconButton aria-label="delete post" onClick={() => deletePost(id)}>
                  <Delete color="secondary" />
                </IconButton>
              </span>
              <Link
                to={`/${extraIds.next}`}
                className={extraIds.next ? 'next' : 'disabled-link next'}
              >
                <ArrowForward fontSize="small" />
              </Link>
            </div>
            <div className="post">
              <span className="author">{post.author}</span>
              <span className="date">
                <small>{post.createdAt}</small>
              </span>
              <h1 className="title">{post.title}</h1>
              <p className="text">{post.text}</p>
            </div>
          </div>
        ) : (
          <Loader />
        )}
        {post && (
          <AddPost
            type="UPDATE"
            open={openEditModal}
            handleClose={() => { this.setState({ openEditModal: false }); this.setPost(); }}
            handleSave={editPost}
            article={post}
          />
        )}
      </div>
    );
  }
}
