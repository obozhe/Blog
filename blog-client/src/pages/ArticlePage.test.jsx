import React from 'react';
import {
  shallow, mount, render, configure,
} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import ArticlePage from './ArticlePage';

configure({ adapter: new Adapter() });

const post = {
  _id: '1',
  author: 'admin',
  title: 'testTitle1',
  text: 'testText1',
  createdAt: '2020-02-18T20:07:54.056Z',
};

const deletePost = () => {};

const editPost = () => {};

const getPost = () => new Promise((res) => res(post));

const formatDate = (d) => {
  const date = new Date(d);
  return `${date.getFullYear()}.${date.getMonth() + 1}.${date.getDate()} ${date.getHours()}:${date.getMinutes()}`;
};

describe('Article component', () => {
  let component;
  let componentInstance;

  beforeEach(() => {
    component = shallow(
      <ArticlePage
        id={post.id}
        extraIds={{ prev: null, next: 2 }}
        deletePost={deletePost}
        editPost={editPost}
        getPost={getPost}
      />,
    );
    componentInstance = component.instance();
  });

  describe('unit tests', () => {
    it('should render without crashes', () => {
      expect(component).toHaveLength(1);
    });

    it('should match snapshot', () => {
      expect(component).toMatchSnapshot();
    });

    it('should disable prev link if prevId is null', () => {
      const prev = component.find('.prev');
      const next = component.find('.next');
      expect(prev.hasClass('disabled-link')).toBeTruthy();
      expect(next.hasClass('disabled-link')).toBeFalsy();
    });

    it('should show title, text, author, date', () => {
      expect(component.find('.author').text()).toEqual(post.author);
      expect(component.find('.date').text()).toEqual(post.createdAt);
      expect(component.find('.title').text()).toEqual(post.title);
      expect(component.find('.text').text()).toEqual(post.text);
    });
  });

  describe('integration tests', () => {
    it('should load post by id', async () => {
      await componentInstance.componentDidMount();
      expect(component.state('post')).toEqual(post);
    });
  });
});
