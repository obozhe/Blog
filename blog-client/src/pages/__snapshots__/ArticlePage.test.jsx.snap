// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article component unit tests should match snapshot 1`] = `ShallowWrapper {}`;
